Canvas.toDataURL causes excessive GPUP memory use on Cocoa

nmoucht opened the pull request:
#### bc684eeb57fb0cb64b1a8d4f05e00dd1cb25bf02
<pre>
Canvas.toDataURL causes excessive GPUP memory use on Cocoa
<a href="https://bugs.webkit.org/show_bug.cgi?id=240639">https://bugs.webkit.org/show_bug.cgi?id=240639</a>
&lt;rdar://93567863&gt;

Reviewed by Said Abou-Hallawa.

From Kimmo&apos;s WIP: #1407

ImageBuffer::toData() and ImageBuffer::toDataURL() would construct a temporary
image-sized bitmap in the GPUP to act as the data encoding source image.
This intermediate bitmap is large for large source images, and will easily make GPUP
memory footprint exceed the limit.

Use the ImageBuffer drawing mechanism to composite the encoding source. This ensures
that the intermediate bitmaps are constructed similar to any other ImageBuffer, and as
such attributes the memory to WP.

Copy the ImageBuffer source to WP instead of GPUP and run the encoding in WP. This makes
the encoding cycles be spent in the correct process and reduces the code footprint of
GPUP.

ImageBuffer::toData() and ::toDataURL() were accessors that were propagated all the way
to the ImageBufferBackend and in GPUP case across processes. They are redundant, as
the data being used is available through copyNativeImage / copyImage.

* LayoutTests/fast/canvas/toDataURL-alpha-permutation-expected.html: Added.
* LayoutTests/fast/canvas/toDataURL-alpha-permutation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::copyNativeImageImpl):
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::copyImage const):
(WebCore::ImageBuffer::toDataURL const):
(WebCore::ImageBuffer::toData const):
(WebCore::ImageBuffer::toDataURL):
(WebCore::ImageBuffer::toData):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.cpp:
(WebCore::ImageBufferCairoBackend::toDataURL const): Deleted.
(WebCore::ImageBufferCairoBackend::toData const): Deleted.
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
(WebCore::imageTypeFromMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encodeImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::createCroppedImageIfNecessary): Deleted.
(WebCore::colorSpaceForBitmap): Deleted.
(WebCore::ImageBufferCGBackend::copyCGImageForEncoding const): Deleted.
(WebCore::ImageBufferCGBackend::toData const): Deleted.
(WebCore::ImageBufferCGBackend::toDataURL const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::copyCGImageForEncoding const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::jpegUTI):
(WebCore::utiFromImageBufferMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encode):
(WebCore::encodeToVector):
(WebCore::encodeToDataURL):
(WebCore::data):
(WebCore::dataURL):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataURLForImageBufferWithQualifiedIdentifier): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBufferWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/API/c/cg/WKImageCG.cpp:
(WKImageCreateDataURLFromImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackendProxy::getDataForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::toDataURL const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::toData const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: <a href="https://commits.webkit.org/251974@main">https://commits.webkit.org/251974@main</a>
</pre>

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -32,25 +32,64 @@
+static RefPtr<NativeImage> copyNativeImageImpl(Ref<ImageBuffer> source, BackingStoreCopy copyBehavior, PreserveResolution preserveResolution)
+{
+    if (source->resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes) {
+        if (source->hasOneRef())
```
How can this be true? It is called like this:

`auto image = copyNativeImageImpl(const_cast<ImageBuffer&>(*this), copyBehavior, preserveResolution);`

This will construct a new Ref<ImageBuffer> and this will increment the refCount by 1. 

It is also called like this from `ImageBuffer::toDataURL()` and `ImageBuffer::toData()`

`image = copyNativeImageImpl(WTFMove(source), DontCopyBackingStore, preserveResolution);`

But the caller of these functions call them like this:

`return toDataURL(Ref { const_cast<ImageBuffer&>(*this) }, mimeType, quality, preserveResolution);`

So when this function is called the refCount is at least equal to 2.

nmoucht replied:
Yeah I'm not sure I understand what this part is trying to do either. Will ping Kimmo to add his thoughts here.

kkinnunen-apple replied:
They're implemented in this patch as  consistency to other logically similar code.

- copy has sinkInto
- draw has draw consuming
- toDataURL has consuming toDataURL 

Many call-sites of `ImageBuffer::toDataURL()` can/will be converted to calls to `static ImageBufferToDataURL(Ref<ImageBuffer>...)` later. Many callers of toDataURL create the image buffer only for the purpose of taking the data url. This change is omitted for simplifying this patch.

Granted, it is a bit out of place after the above described change was removed for simplicity. I guess the static functions can be removed and added in a later patch if they're bothersome.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -32,25 +32,64 @@
 static const float MaxClampedLength = 4096;
 static const float MaxClampedArea = MaxClampedLength * MaxClampedLength;
 
+static RefPtr<ImageBuffer> copyImageBuffer(Ref<ImageBuffer> source, PreserveResolution preserveResolution)
```
Can we implement ImageBuffer::clone() by using this function? Also can we use the same naming: ImageBuffer::clone() and cloneImageBuffer() or ImageBuffer::copy() and copyImageBuffer()?

nmoucht replied:
Use copyImageBuffer() in clone() now.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -32,25 +32,64 @@
 #include "PlatformImageBuffer.h"
+#include <wtf/text/Base64.h>
+
+#if USE(CG)
+#include "ImageBufferUtilitiesCG.h"
+#endif
+#if USE(CAIRO)
+#include "ImageBufferUtilitiesCairo.h"
+#endif
```
It is strange that we use platform headers and there is no platform code specific in this file. I think these header files are included because of using platformMIMETypeIsJPEG(). If this is the case I do not think we need to repeat twice in two different header files.

nmoucht replied:
These are necessary as data (not named encodeData) is platform specific.

shallawa replied:
We usually have one prototype for the function and multiple implementations for it. So I would suggest the following since you are modifying this code:

1. Rename Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.h to Source/WebCore/platform/graphics/ImageBufferUtilities.h
2. Remove the prototype of encodeData() from from ImageBufferUtilitiesCG.h
3. Include ImageBufferUtilities.h here and every place which calls encodeData().

nmoucht replied:
This is a good idea. I'll do this in a separate patch.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -171,4 +201,41 @@ void ImageBuffer::drawConsuming(RefPtr<ImageBuffer> imageBuffer, GraphicsContext
+            return { };
+        // We composite the copy on top of black by drawing black under the copy.
+        copyBuffer->context().fillRect({ { }, copyBuffer->logicalSize() }, Color::black, CompositeOperator::DestinationOver);
+        image = sinkIntoNativeImage(WTFMove(copyBuffer));
```
Should this code be moved to another static function like copyOpaqueNativeImageImpl() or something similar?

nmoucht replied:
Moved to copyOpaqueNativeImageImpl().

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -171,4 +201,41 @@ void ImageBuffer::drawConsuming(RefPtr<ImageBuffer> imageBuffer, GraphicsContext
+    }
+    if (!image)
+        return { };
+    return data(image->platformImage().get(), mimeType, quality);
```
Not new. But data() is a little bit generic name and does not reveal what it is doing. Can we use a better name like encodeToData() or something similar?

nmoucht replied:
Changed to encodeData().

shallawa commented on Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
```diff
@@ -50,6 +50,17 @@
+{
+    return imageTypeFromMIMEType(mimeType) == "jpeg"_s;
+
+}
```
These function do not belong to this file. I think they should be moved to MIMETypeRegistry.cpp. See MIMETypeRegistry::isTextMIMEType() and isPostScriptMIMEType() as examples.

nmoucht replied:
Moved to MIMETypeRegistry.cpp as isJPEGMIMEType.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -156,6 +156,8 @@ unsigned ImageBufferIOSurfaceBackend::bytesPerRow() const
 
 void ImageBufferIOSurfaceBackend::invalidateCachedNativeImage() const
 {
+    if (!m_mayHaveOutstandingBackingStoreReferences)
+        return;
```
Is this change relevant to this patch?

nmoucht replied:
I believe this was added due to another change.

shallawa replied:
I see. I think it is an optimization. So keeping it is fine.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -201,9 +194,9 @@ RefPtr<PixelBuffer> ImageBufferIOSurfaceBackend::getPixelBuffer(const PixelBuffe
 
 void ImageBufferIOSurfaceBackend::putPixelBuffer(const PixelBuffer& pixelBuffer, const IntRect& srcRect, const IntPoint& destPoint, AlphaPremultiplication destFormat)
 {
+    invalidateCachedNativeImage();
```
Is this change relevant to this patch?

nmoucht replied:
Same as above.

shallawa replied:
Yes I think we need this change. If we are going to change the IOSurface pixels directly without going through the CG, we need to tell CG to copy the backing store of the copied images which were created by copy-on-write now before we call `putPixelBuffer()`.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
```diff
@@ -96,9 +96,21 @@ ALLOW_DEPRECATED_DECLARATIONS_BEGIN
 }
 
-static bool encode(CGImageRef image, CFStringRef destinationUTI, std::optional<double> quality, const ScopedLambda<PutBytesCallback>& function)
+bool platformMIMETypeIsJPEG(const String& mimeType)
```
I do not think we need different platform implementations for this function. I think we can just use one function named MIMETypeRegistry::isJPEGMIMEType() for example.

nmoucht replied:
Moved to MIMETypeRegistry.cpp as isJPEGMIMEType.

shallawa commented on LayoutTests/fast/canvas/toDataURL-alpha-permutation.html:
```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE html> 
```
Does this test fail without this patch? Does this patch fix a correctness issue or it is memory and perf optimization?

nmoucht replied:
This test was added due to the removal of kCGImageAlphaNoneSkipLast. This patch is primarily a memory optimization, but we want to make sure it still works for different alpha values.

shallawa commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -832,4 +833,19 @@ Vector<String> MIMETypeRegistry::allowedFileExtensions(const Vector<String>& mim
+    return mimeType.substring(sizeof "image") == "jpeg"_s;
+#endif
+    ASSERT_NOT_REACHED();
+    return false;
```
I do not know why we have to have a special case for CG. But let's leave it as is for now. However I think we can have string matching code run for all not CG ports:

```
#if USE(CG)
    ...
#else
    return mimeType.substring(sizeof "image") == "jpeg"_s;
#endif

```

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -31,26 +31,77 @@
+    return copyBuffer;
+}
+
+static RefPtr<NativeImage> copyNativeImageImpl(Ref<ImageBuffer> source, BackingStoreCopy copyBehavior, PreserveResolution preserveResolution)
```
I think it is not clear from this name what this function is doing. The postfix "Impl" does not reveal the difference between `ImageBuffer::copyNativeImage()` and `copyNativeImageImpl()`.

 So I would suggesting the name `copyImageBufferToNativeImage()` or something similar.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -31,26 +31,77 @@
+    return ImageBuffer::sinkIntoNativeImage(WTFMove(copyBuffer));
+}
+
+static RefPtr<NativeImage> copyOpaqueNativeImageImpl(Ref<ImageBuffer> source, PreserveResolution preserveResolution)
```
`copyImageBufferToOpaqueNativeImage()` or something similar .

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -171,4 +208,30 @@ void ImageBuffer::drawConsuming(RefPtr<ImageBuffer> imageBuffer, GraphicsContext
+    if (!image)
+        return { };
+    return encodeData(image->platformImage().get(), mimeType, quality);
+}
```
Why are these function added as static of ImageBuffer? Are they really needed? I do not see them called in this patch expect from the member functions toDataURL() and toData().

Also this pattern is different from the pattern above where copyNativeImageImpl() is called from ImageBuffer::copyImage() and where copyImageBuffer() is called from ImageBuffer::clone().

nmoucht replied:
I believe these are relevant to the next step after this patch that Kimmo mentioned above.

nmoucht committed:
Canvas.toDataURL causes excessive GPUP memory use on Cocoa
https://bugs.webkit.org/show_bug.cgi?id=240639
<rdar://93567863>

Reviewed by Said Abou-Hallawa.

From Kimmo's WIP: #1407

ImageBuffer::toData() and ImageBuffer::toDataURL() would construct a temporary
image-sized bitmap in the GPUP to act as the data encoding source image.
This intermediate bitmap is large for large source images, and will easily make GPUP
memory footprint exceed the limit.

Use the ImageBuffer drawing mechanism to composite the encoding source. This ensures
that the intermediate bitmaps are constructed similar to any other ImageBuffer, and as
such attributes the memory to WP.

Copy the ImageBuffer source to WP instead of GPUP and run the encoding in WP. This makes
the encoding cycles be spent in the correct process and reduces the code footprint of
GPUP.

ImageBuffer::toData() and ::toDataURL() were accessors that were propagated all the way
to the ImageBufferBackend and in GPUP case across processes. They are redundant, as
the data being used is available through copyNativeImage / copyImage.

* LayoutTests/fast/canvas/toDataURL-alpha-permutation-expected.html: Added.
* LayoutTests/fast/canvas/toDataURL-alpha-permutation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::copyNativeImageImpl):
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::copyImage const):
(WebCore::ImageBuffer::toDataURL const):
(WebCore::ImageBuffer::toData const):
(WebCore::ImageBuffer::toDataURL):
(WebCore::ImageBuffer::toData):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.cpp:
(WebCore::ImageBufferCairoBackend::toDataURL const): Deleted.
(WebCore::ImageBufferCairoBackend::toData const): Deleted.
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
(WebCore::imageTypeFromMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encodeImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::createCroppedImageIfNecessary): Deleted.
(WebCore::colorSpaceForBitmap): Deleted.
(WebCore::ImageBufferCGBackend::copyCGImageForEncoding const): Deleted.
(WebCore::ImageBufferCGBackend::toData const): Deleted.
(WebCore::ImageBufferCGBackend::toDataURL const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::copyCGImageForEncoding const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::jpegUTI):
(WebCore::utiFromImageBufferMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encode):
(WebCore::encodeToVector):
(WebCore::encodeToDataURL):
(WebCore::data):
(WebCore::dataURL):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataURLForImageBufferWithQualifiedIdentifier): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBufferWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/API/c/cg/WKImageCG.cpp:
(WKImageCreateDataURLFromImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackendProxy::getDataForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::toDataURL const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::toData const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: https://commits.webkit.org/251974@main

csaavedra commented on Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
```diff
@@ -78,31 +78,11 @@ static bool encodeImage(cairo_surface_t* surface, const String& mimeType, std::o
     // http://developer.gnome.org/gdk-pixbuf/stable/gdk-pixbuf-File-saving.html#gdk-pixbuf-save-to-bufferv
 
-    String type = mimeType.substring(sizeof "image");
+    String type = imageTypeFromMIMEType(mimeType);
```
This change broke the GTK+ build. There is no `imageTypeFromMIMEType()` defined anywhere, but I see it mentioned in the commit log. Did you forget to add the chunk to the commit?

csaavedra commented on Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
```diff
@@ -65,7 +65,7 @@ static bool encodeImage(cairo_surface_t* image, const String& mimeType, Vector<u
 }
 
-Vector<uint8_t> data(cairo_surface_t* image, const String& mimeType, std::optional<double>)
+Vector<uint8_t> encodeData(cairo_surface_t* image, const String& mimeType, std::optional<double>)
```
This definition is inside a `#if !PLATFORM(GTK)` but the method is also needed in the GTK port, so this breaks the build as well.
